src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+        /// </summary>
+        /// <param name="attrName">
+        /// attribute names need to be fully qualified, 
+        /// e.g. DefaultOverload is really Windows.Foundation.Metadata.DefaultOverload
+        /// </param>
+        /// <param name="ls">all the syntax nodes that correspond to an attribute list</param>
+        /// <returns>true iff the given attribute is in the list</returns>
+        private bool MatchesAnyAttribute(string attrName, SyntaxList<AttributeListSyntax> ls)
+        {
+            foreach (var attrList in ls)
+            {
+                foreach (var attr in attrList.Attributes)
+                {
+                    if (attr.Name.ToString().Equals(attrName))
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Looks at all possible attributes on a given method declaration 
+        /// </summary>
+        /// <returns>
+        /// returns true iff any are (string) equal to the given attribute name 
+        /// </returns>
+        private bool MethodHasAttribute(string attrName, MethodDeclarationSyntax method) { return MatchesAnyAttribute(attrName, method.AttributeLists); }
+
+        /// <summary>
+        /// Looks at all possible attributes on a given parameter declaration 
+        /// </summary>
+        /// <returns>
+        /// returns true iff any are (string) equal to the given attribute name 
+        /// </returns>
+        private bool ParamHasAttribute(string attrName, ParameterSyntax param) { return MatchesAnyAttribute(attrName, param.AttributeLists); }
+
+        /// <summary>
+        /// Keeps track of repeated declarations of a method (overloads) and raises diagnostics according to the rule that exactly one overload should be attributed the default
+        /// </summary>
+        /// <param name="method">
+        /// The method to check attributes for</param>
+        /// <param name="methodHasAttributeMap">
+        /// Keeps track of the method (via name + arity) and whether it was declared with the DefaultOverload attribute</param>
+        /// <param name="overloadsWithoutAttributeMap">
+        ///     Keeps track of the methods that are overloads but don't have the DefaultOverload attribute (yet)
+        ///     Used after this function executes, hence the reference parameter 
+        ///     </param>
+        /// <param name="classId">The class the method lives in -- used for creating the diagnostic</param>
+        /// <param name="context">The SourceGenerator context the code lives in</param>
+        /// <returns>True iff multiple overloads of a method are found, where more than one has been designated as the default overload</returns>
+        private bool CheckOverloadAttributes(MethodDeclarationSyntax method,
+            ref Dictionary<string, bool> methodHasAttributeMap,

```

Comment:
Any reason to pass these by ref?  https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref#passing-an-argument-by-reference-an-example

In C#, objects are by default passed by reference. I believe ref here is only helpful if you plan on changing the object value itself here.

Reply:
When I started out, I wasn't passing by ref and diagnostics weren't sticking to the context, so I wanted to change the context value to have the diagnostic I reported. I'll see what happens if I take `ref` out again

Reply:
Taking them all out only causes 42 (out of 257) tests to fail, I'll investigate and see if we can get by without `ref`

Reply:
I have a second PR for after this one and I have it working with no explicit `ref` params